options: Add kong.ShortUsageOnError() option

juliaogris opened the pull request:
Add `kong.ShortUsageOnError()` option similar to `kong.UsageOnError()`.
For a custom short usage message use `kong.ShortHelp(kong.HelpPrinter)`
Add tests for UsageOnError and ShortUsageOnError.

juliaogris commented on help.go:
```diff
@@ -115,6 +112,9 @@ func printApp(w *helpWriter, app *Application) {
 		}
+	} else if w.Summary && app.HelpFlag != nil {
+		w.Print("")
+		w.Printf(`Run "%s --help" for more information.`, app.Name)
```
Not sure if this is OK.

juliaogris commented on help.go:
```diff
@@ -89,9 +89,6 @@ func DefaultHelpValueFormatter(value *Value) string {
 func DefaultHelpPrinter(options HelpOptions, ctx *Context) error {
-	if ctx.Empty() {
-		options.Summary = false
-	}
```
Not sure if this is OK.

alecthomas replied:
Why remove it, was it doing something unexpected?

juliaogris replied:
The usage message doesn't get printed as short (but long/full) is I keep these lines.
`helopoOtions.Summary` is how indicate that I want the short usage - maybe that was a bad choice?
 `options.Summary = true` is set in under `ShortUsageOnError` in `kong.go` and then reset here.

alecthomas commented on kong.go:
```diff
@@ -47,14 +47,15 @@ type Kong struct {
-	groups        []Group
-	vars          Vars
+	noDefaultHelp     bool
+	usageOnError      bool
```
I think either combine these two variables into an enum (none, full, short), or ensure that both of them can't be set at once in the options, otherwise it could be confusing.

juliaogris replied:
Fair enough - I'm not too sure which one to go for:
```
kong.UsageOnError(kong.Full)
kong.UsageOnError(kong.Short)
kong.UsageOnError(kong.None)
```
is a bit of a mouth full and really it should probably be `kong.UsageOnError(kong.UsageNone)` etc., right? So even longer.

However, I don't see where the best place would be for option validation to only allow one of `kong.UsageOnError` or `kong.ShortUsageOnError`.

What do you suggest?

alecthomas replied:
We can't change the interface, I more meant that the two functions should ensure that the other isn't set, by unsetting the other's state or changing the internal values to a single enum (not exposing it).

alecthomas commented on help_test.go:
```diff
@@ -512,3 +512,55 @@ Group 2
+	)
+	_, err := p.Parse([]string{})
+	require.Error(t, err)
+	p.FatalIfErrorf(err)
```
Remove?

juliaogris replied:
It doesn't print without it (actual output is empty, just double checked).
I followed the flow from `Parse` in `global.go` - is there a better way of doing this?

alecthomas commented on help_test.go:
```diff
@@ -512,3 +512,55 @@ Group 2
+		kong.Exit(func(int) {}),
+		kong.UsageOnError(),
+	)
+	_, err := p.Parse([]string{"--help"})
```
Shouldn't this be parsing an empty slice to test the error?

juliaogris replied:
Yes, sorry. Sloppy. Fixed in fixup commit 891ebfd .

alecthomas commented on help_test.go:
```diff
@@ -512,3 +512,55 @@ Group 2
+	require.Error(t, err)
+	p.FatalIfErrorf(err)
+
+	expected := `Usage: test --flag=STRING
```
Is this the kind of help you're aiming for? It's not as succinct as I was imagining! Feel free to adjust the output further if you prefer.

juliaogris replied:
I am not sure. What were you thinking? Without description it would be:
```
Usage: test --flag=STRING

Run "test --help" for more information.

test: error: missing flags: --flag=STRING
```
I think I would prefer:

```
Usage: test --flag=STRING
Run "test --help" for more information.

test: error: missing flags: --flag=STRING
```
but I'm not very opinionated here.

The reason I left the description in apart from it being the normal flow anyway is that I use the description together with `NoAppSummary: true` to get a custom Usage line highlighting the flags I want. I suppose that's a bit of a hack in the first place.

Do you think I should just add another HelpOption field called say `Short` and create a custom output for it?

alecthomas replied:
I honestly can't remember what we were trying to achieve! :) I think it was motivated by something you did in gurl? Let's have a chat about it on Monday?

juliaogris reviewed (commented):
I've responded to all comments.
Is the general direction of this PR ok? 
Should I use `HelpOptions.Summary` to mark usage output as short/long from `FatalIfErr` or should I introduce another HelpOption field or do something completely different?

Thanks.

juliaogris commented:
@alecthomas - I've given it another pass; I've added the usageOnError enum as suggested and I've moved away from (ab)using `HelpOptions.Summary` as an indicator for `Short|FullUsage` message and instead introduced a `ShortHelpOption` and `DefaultShortHelpPrinter`. It is still a little awkward because the `ShortHelp` option doesn't really make a lot of sense without `ShortUsageOnError` - but maybe this is closer to acceptable?

alecthomas commented on options.go:
```diff
@@ -192,6 +192,14 @@ func Help(help HelpPrinter) Option {
 	})
 }
 
+// ShortHelp printer to use.
```
Can you expand this comment to explain the intent of this and when it would be used?

alecthomas commented on options.go:
```diff
@@ -251,7 +259,15 @@ func ExplicitGroups(groups []Group) Option {
+	})
+}
+
+// ShortUsageOnError configures Kong to display context-sensitive short usage if FatalIfErrorf is called with an error.
```
Please add an explanation that this will trigger `ShortHelp()` formatting.

alecthomas reviewed (approved):
Minor comments but LGTM!

juliaogris commented:
I've added changes according to last comments and squashed the fixup-commits into a single commit.
